File: android/app/src/main/kotlin/com/kthakrebts/flutter_ble_architecture/BleManager.kt
package com.kthakrebts.flutter_ble_architecture

import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Main coordinator bridging Flutter MethodChannel to Android Native BLE APIs.
 */
class BleManager(private val context: Context, messenger: BinaryMessenger) : MethodChannel.MethodCallHandler {

    private val TAG = "BleManager"
    private val channel = MethodChannel(messenger, "com.kthakrebts.ble/bridge")

    private val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val bluetoothAdapter = bluetoothManager.adapter

    private lateinit var scanner: BleScanner
    private lateinit var connectionManager: BleConnectionManager

    init {
        channel.setMethodCallHandler(this)
        
        if (bluetoothAdapter != null) {
            // Scanner callback
            scanner = BleScanner(context, bluetoothAdapter) { scanResult ->
                val deviceMap = mapOf(
                    "name" to (scanResult.device.name ?: "Unknown Device"),
                    "address" to scanResult.device.address,
                    "rssi" to scanResult.rssi
                )
                // Stream scan results back to Flutter UI asynchronously
                runOnMainThread {
                    channel.invokeMethod("onDeviceDiscovered", deviceMap)
                }
            }

            connectionManager = BleConnectionManager(context, bluetoothAdapter)
        } else {
            Log.e(TAG, "Bluetooth Adapter not available on this hardware.")
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (bluetoothAdapter == null) {
            result.error("BLE_UNAVAILABLE", "Bluetooth hardware is not supported on this device.", null)
            return
        }

        when (call.method) {
            "startScan" -> {
                try {
                    scanner.startScan()
                    result.success(true)
                } catch (e: Exception) {
                    result.error("SCAN_ERROR", e.message, null)
                }
            }
            "stopScan" -> {
                try {
                    scanner.stopScan()
                    result.success(true)
                } catch (e: Exception) {
                    result.error("SCAN_ERROR", e.message, null)
                }
            }
            "connect" -> {
                val address = call.argument<String>("address")
                if (address == null) {
                    result.error("INVALID_ARGUMENT", "MAC address is required.", null)
                    return
                }

                val callback = GattCallbackHandler(
                    onConnectionStateChanged = { devAddress, status, newState ->
                        val stateMap = mapOf(
                            "address" to devAddress,
                            "status" to status,
                            "state" to newState
                        )
                        runOnMainThread {
                            channel.invokeMethod("onConnectionStateChanged", stateMap)
                        }
                    },
                    onServicesDiscovered = { devAddress, status ->
                        val serviceList = connectionManager.getGatt(devAddress)
                            ?.services
                            ?.map { it.uuid.toString() } ?: emptyList()
                        
                        val serviceMap = mapOf(
                            "address" to devAddress,
                            "status" to status,
                            "services" to serviceList
                        )
                        runOnMainThread {
                            channel.invokeMethod("onServicesDiscovered", serviceMap)
                        }
                    },
                    onCharacteristicRead = { devAddress, charUuid, value, status ->
                        val charMap = mapOf(
                            "address" to devAddress,
                            "characteristic" to charUuid,
                            "value" to value,
                            "status" to status
                        )
                        runOnMainThread {
                            channel.invokeMethod("onCharacteristicRead", charMap)
                        }
                    }
                )

                val success = connectionManager.connect(address, callback)
                result.success(success)
            }
            "disconnect" -> {
                val address = call.argument<String>("address")
                if (address == null) {
                    result.error("INVALID_ARGUMENT", "MAC address is required.", null)
                    return
                }
                connectionManager.disconnect(address)
                result.success(true)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun runOnMainThread(action: () -> Unit) {
        val mainHandler = android.os.Handler(context.mainLooper)
        mainHandler.post(action)
    }

    fun onDestroy() {
        if (::connectionManager.isInitialized) {
            connectionManager.disconnectAll()
        }
    }
}

File: ios/Runner/BLEManager.swift
import CoreBluetooth
import Flutter
import Foundation

/// Main bridge coordinator between Flutter and iOS CoreBluetooth framework.
class BLEManager: NSObject, CBCentralManagerDelegate {
    private var centralManager: CBCentralManager?
    private var channel: FlutterMethodChannel?
    
    private var scanner: BLEScanner?
    private var connectionManager: BLEConnectionManager?
    
    // Scanned but not necessarily connected peripherals dictionary to look up for connection
    private var discoveredPeripherals = [String: CBPeripheral]()

    init(messenger: FlutterBinaryMessenger) {
        super.init()
        self.channel = FlutterMethodChannel(name: "com.kthakrebts.ble/bridge", binaryMessenger: messenger)
        self.centralManager = CBCentralManager(delegate: self, queue: nil)
        
        self.channel?.setMethodCallHandler { [weak self] (call, result) in
            self?.handle(call, result: result)
        }
        
        // Initialize sub-managers
        self.scanner = BLEScanner(centralManager: centralManager) { [weak self] deviceMap in
            self?.channel?.invokeMethod("onDeviceDiscovered", arguments: deviceMap)
        }
        
        self.connectionManager = BLEConnectionManager(
            centralManager: centralManager,
            onConnectionChanged: { [weak self] uuid, state in
                let stateMap: [String: Any] = ["address": uuid, "state": state]
                self?.channel?.invokeMethod("onConnectionStateChanged", arguments: stateMap)
            },
            onServicesDiscovered: { [weak self] uuid, services in
                let serviceMap: [String: Any] = [
                    "address": uuid,
                    "services": services,
                    "status": 0 // 0 means Success on iOS / Android parity
                ]
                self?.channel?.invokeMethod("onServicesDiscovered", arguments: serviceMap)
            }
        )
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard centralManager?.state == .poweredOn else {
            result(FlutterError(code: "BLE_POWERED_OFF", message: "Bluetooth is not turned on.", details: nil))
            return
        }

        switch call.method {
        case "startScan":
            scanner?.startScan()
            result(true)
        case "stopScan":
            scanner?.stopScan()
            result(true)
        case "connect":
            guard let args = call.arguments as? [String: Any],
                  let uuidStr = args["address"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "MAC/UUID is required.", details: nil))
                return
            }
            
            if let peripheral = discoveredPeripherals[uuidStr] {
                connectionManager?.connect(peripheral: peripheral)
                result(true)
            } else {
                // Try retrieving system connected peripherals or UUID match
                if let uuid = UUID(uuidString: uuidStr),
                   let peripheral = centralManager?.retrievePeripherals(withIdentifiers: [uuid]).first {
                    discoveredPeripherals[uuidStr] = peripheral
                    connectionManager?.connect(peripheral: peripheral)
                    result(true)
                } else {
                    result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device not found in scanned list.", details: nil))
                }
            }
        case "disconnect":
            guard let args = call.arguments as? [String: Any],
                  let uuidStr = args["address"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "MAC/UUID is required.", details: nil))
                return
            }
            connectionManager?.disconnect(uuidString: uuidStr)
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - CBCentralManagerDelegate Callbacks

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        print("BLE Manager: Central manager state updated: \(central.state.rawValue)")
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        let uuidStr = peripheral.identifier.uuidString
        discoveredPeripherals[uuidStr] = peripheral
        scanner?.didDiscoverPeripheral(peripheral, advertisementData: advertisementData, rssi: RSSI)
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        let uuidStr = peripheral.identifier.uuidString
        print("BLE Manager: Connected to \(uuidStr)")
        // State codes: 2 = Connected (Android parity), 0 = Disconnected, 1 = Connecting, 3 = Disconnecting
        let stateMap: [String: Any] = ["address": uuidStr, "state": 2]
        channel?.invokeMethod("onConnectionStateChanged", arguments: stateMap)
        
        // Discover services immediately
        peripheral.discoverServices(nil)
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        let uuidStr = peripheral.identifier.uuidString
        print("BLE Manager: Failed to connect to \(uuidStr)")
        let stateMap: [String: Any] = ["address": uuidStr, "state": 0]
        channel?.invokeMethod("onConnectionStateChanged", arguments: stateMap)
        connectionManager?.removePeripheral(uuidString: uuidStr)
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        let uuidStr = peripheral.identifier.uuidString
        print("BLE Manager: Disconnected from \(uuidStr)")
        let stateMap: [String: Any] = ["address": uuidStr, "state": 0]
        channel?.invokeMethod("onConnectionStateChanged", arguments: stateMap)
        connectionManager?.removePeripheral(uuidString: uuidStr)
    }
}
